syntax = "proto3";

package blog;

service BlogService {
    rpc CreateBlog(CreateBlogRequest) returns (BlogResponse);
    rpc CreatePost(CreatePostRequest) returns (PostResponse);
    rpc PostComment(PostCommentRequest) returns (CommentResponse);
    rpc EditComment(EditCommentRequest) returns (CommentResponse);
    rpc GetAllPosts(Empty) returns (PostsResponse);
    rpc GetAllBlogs(Empty) returns (BlogsResponse);
    rpc GetBlogById(BlogIdRequest) returns (BlogResponse);
    rpc GetPostById(PostIdRequest) returns (PostResponse);
    rpc GetCommentsByPostId(PostIdRequest) returns (CommentsResponse);
}

message Empty {}

message CreateBlogRequest {
    string token = 1;
    string title = 2;
    string description = 3;
}

message BlogResponse {
    string message = 1;
    Blog blog = 2;
}

message CreatePostRequest {
    string token = 1;
    int32 blog_id = 2;
    string title = 3;
    string description = 4;
    string date = 5;
    repeated string images = 6;
    string status = 7;
}

message PostResponse {
    string message = 1;
    Post post = 2;
}

message PostCommentRequest {
    string token = 1;
    int32 post_id = 2;
    string comment_text = 3;
}

message CommentResponse {
    string message = 1;
    Comment comment = 2;
}

message EditCommentRequest {
    string token = 1;
    int32 comment_id = 2;
    string comment_text = 3;
}

message BlogIdRequest {
    int32 blog_id = 1;
}

message PostIdRequest {
    int32 post_id = 1;
}

message BlogsResponse {
    repeated Blog blogs = 1;
}

message PostsResponse {
    repeated Post posts = 1;
}

message CommentsResponse {
    repeated Comment comments = 1;
}

message Blog {
    int32 id = 1;
    int32 user_id = 2;
    string title = 3;
    string description = 4;
    string created_at = 5;
    string author = 6;
}

message Post {
    int32 id = 1;
    int32 blog_id = 2;
    string title = 3;
    string description = 4;
    repeated string images = 5;
    string status = 6;
    string created_at = 7;
}

message Comment {
    int32 comment_id = 1;
    int32 parent_comment_id = 2;
    int32 user_id = 3;
    string comment_text = 4;
    int32 likes = 5;
    string created_at = 6;
    string author = 7;
    repeated Comment replies = 8;
}
